### PR TITLE
[FLINK-24349]Support customized Calalogs via JDBC

### PR DESCRIPTION
**Pull request by cuibo01**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (Airblader) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: We should move all of these parameters into a `Context` interface and pass that so we can easily add more information if we need to without breaking the API. This is similar to how table factories work, and basically the JDBC dialect infrastructure is just a clone of table factories anyway.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (Airblader) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:148]: Logically this seems misplaced to me, because creating a catalog doesn't relate to the definition of a dialect. However, I don't have a better suggestion either, we shouldn't introduce even more custom infrastructure.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (Airblader) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:156]: This load call is unnecessary, it'll just find itself (`this`) anyway. We're already in the loaded dialect here. We can also just directly use `dialectName`.

```suggestion
                String.format("Catalog for JDBC dialect '%s' is not supported yet.", dialectName()));
```

**Review by Airblader — CHANGES_REQUESTED** [review]
Thanks for your PR. I left some comments, and I also have one bigger concern. We're now making `AbstractJdbcCatalog` a public API, so first off, we need to add `@PublicEvolving` to the class.

However, that means we need to ready that class for public usage. It needs proper JavaDoc on the class that describes its purpose and what implementors need to do.  It should also describe how the JDBC catalog works a bit.

We also need to consider the design of the class. At the moment it is hardwired to use username/password for authentication. That isn't very flexible or extensible.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:148]: thx for your review... 
yes, u are right. If we add new interface, we need to modify a lot code. I think the current way might be better than modifying a lot of code.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: u mean we should revise from here `https://github.com/apache/flink/blob/d753cf20826c66f20bfbd3772c58709e00d9bcb7/flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java#L72` ?

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: Yeah, I think passing the original `CatalogFactory#Context` would be best as that gives implementations access to everything a normal catalog factory would have access to, like the configuration and classloader. We should also annotate `JdbcCatalogFactoryOptions` as `@PublicEvolving` then so implementations can use it to retrieve catalog options.

I'm now seeing that `JdbcCatalog` is already public and hard-wired to use username/password. That is unfortunate.

Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: ok，i prefer to create new issue for `Context`. of course, if u think it is better to update `Context` in the patch, i can modify it

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: Since we are creating public API in this PR, we need to make sure the API is sound in this PR and not defer it to another task.

Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: but the JdbcDialect already has `@PublicEvolving`, https://github.com/apache/flink/blob/d753cf20826c66f20bfbd3772c58709e00d9bcb7/flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java#L35

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: But you are adding a new public API method to this interface in this PR.

Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: > But you are adding a new public API method to this interface in this PR. So we need to make sure that the design of this method is good.

sorry, i dont understand how I should do it... Can you go into more detail?

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: Basically, I think we should just pass the `context` argument from `JdbcCatalogFactory#createCatalog` directly to `JdbcDialect#createCatalog` instead of "destructuring" it into into a list of the individual options. Then any changes to `CatalogFactory#Context` would also automatically become available for dialect implementations.

I think we should also remove `username`, `pwd`, `baseUrl`, `defaultUrl`, `getUsername`, `getPassword`, and `getBaseUrl` from `AbstractJdbcCatalog`. We can either move the connection test logic of `#open` into `JdbcCatalogUtils` and call it from `PostgresCatalog`, or we instead add a `abstract Connection getConnection()` method to `AbstractJdbcCatalog` and use that in `#open`. This decouples the catalog implementation from the concrete authentication implementation.

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: Maybe CC @bowenli86 and @JingsongLi to give feedback on this, too.

Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: > Basically, I think we should just pass the `context` argument from `JdbcCatalogFactory#createCatalog` directly to `JdbcDialect#createCatalog` instead of "destructuring" it into into a list of the individual options. Then any changes to `CatalogFactory#Context` would also automatically become available for dialect implementations.
> 
> I think we should also remove `username`, `pwd`, `baseUrl`, `defaultUrl`, `getUsername`, `getPassword`, and `getBaseUrl` from `AbstractJdbcCatalog`. Instead add a `abstract Connection getConnection()` method to `AbstractJdbcCatalog` and use that in `#open`. This decouples the catalog implementation from the concrete authentication implementation.
> 
> But I'm happy to hear other people's thoughts on this, too.

thx @Airblader  got it. but i prefer to create a new refactored issue

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: I'd be OK with doing this in a separate ticket, but then that ticket needs to be done first before we can merge this. This creates new public APIs, and public APIs need to be correct when they are merged. Otherwise the other ticket will fall through the cracks and this gets released as-is. I'm not a committer, though, this is just my opinion.

Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: > Otherwise the other ticket will fall through the cracks and this gets released as-is. I'm not a committer, though, this is just my opinion.

i am sorry, i dont understand why can not we merge the PR first and then create a new ticket?

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: Once a new public API is released, we cannot just change it anymore – because it is a breaking change for users, and we need to avoid breaking changes. This is different from internal implementations which we can change whenever we need to. In this case, the internal implementation doesn't make for a good public API, so we need to bring it into better shape. If we just open another issue, we cannot ensure that this issue would be resolved prior to the next release.

**Comment by cuibo01** [comment]
WDYT? @bowenli86 @JingsongLi

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: The [current discussion](https://lists.apache.org/x/thread.html/ra438e89eeec2d4623a32822e21739c8f2229505522d73d1034e34198@%3Cdev.flink.apache.org%3E) on the dev ML really emphasizes that we simply need to put more focus and effort into designing APIs well and avoid breaking them.

Comment (cuibo01) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: Yes, you are right. 
But I dont understand why you have to refactor the code on this PR. This PR is designed to support custom calalogs :-)

Comment (Airblader) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:149]: Because AbstractJdbcCatalog is currently not a public API, but with your PR it becomes one. Therefore it must have a good design or else we'll be forced to break the API in the future, which is exactly what we need to avoid. I cannot think of more ways to say this. Why do you want to move this to a separate issue? It only creates a big risk of releasing a bad API because no one implements the follow-up issue. Things should have the proper quality when they hit master, not when a release is created.

**Comment by MartijnVisser** [comment]
@cuibo01 Is there any progress to report from your end?

**Comment by MartijnVisser** [comment]
@cuibo01 Closing due to lack of inactivity. If you're still open for this idea, please create a new PR in the dedicated JDBC connector repository

**Pull request closed without merging**
